source/cgal/polygon.cc

```diff
@@ -0,0 +1,171 @@
+      {
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[0]));
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[1]));
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[2]));
+      }
+    else if (cell->reference_cell() == ReferenceCells::Quadrilateral)
+      {
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[0]));
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[1]));
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[3]));
+        polygon.push_back(
+          CGALWrappers::dealii_point_to_cgal_point<CGAL::Point_2<KernelType>,
+                                                   2>(vertices[2]));
+      }
+    else
+      {
+        DEAL_II_ASSERT_UNREACHABLE();
+      }
+  }
+
+
+
+  template <typename KernelType>
+  void
+  dealii_tria_to_cgal_polygon(const Triangulation<2, 2>   &tria,
+                              CGAL::Polygon_2<KernelType> &polygon)
+  {
+    std::map<unsigned int, unsigned int> face_vertex_indices;
+
+    for (const auto &cell : tria.active_cell_iterators())
+      {
+        for (const unsigned int i : cell->face_indices())
+          {
+            const typename Triangulation<2, 2>::face_iterator &face =
+              cell->face(i);
+            if (face->at_boundary())
+              {
+                if (cell->reference_cell() == ReferenceCells::Quadrilateral &&
+                    (i == 0 || i == 3))
+                  {
+                    face_vertex_indices[face->vertex_index(1)] =
+                      face->vertex_index(0);
+                  }
+                else
+                  {
+                    face_vertex_indices[face->vertex_index(0)] =
+                      face->vertex_index(1);
+                  }
+              }
+          }
+      }
+
+    const auto &vertices = tria.get_vertices();
+    polygon.clear();
+
+    unsigned int current_index = face_vertex_indices.begin()->first;
+
+    for (size_t i = face_vertex_indices.size(); i > 0; --i)
+      {
+        polygon.push_back(
+          dealii_point_to_cgal_point<CGAL::Point_2<KernelType>, 2>(
+            vertices[current_index]));
+        auto it       = face_vertex_indices.find(current_index);
+        current_index = it->second;
+        face_vertex_indices.erase(it);
+      }
```

Comment:
At the very end of the loop, you should have gotten back to the starting index. If you didn't, you made a mistake, but that is something you could check with an assertion.